pkg/volume/csi/csi_client.go

```diff
@@ -66,66 +68,172 @@ type csiClient interface {
-	NodeGetCapabilities(ctx context.Context) ([]*csipb.NodeServiceCapability, error)
+	NodeSupportsStageUnstage(ctx context.Context) (bool, error)
 }
 
+// Strongly typed address
+type csiAddr string
+
+// Strongly typed driver name
+type csiDriverName string
+
 // csiClient encapsulates all csi-plugin methods
 type csiDriverClient struct {
-	driverName        string
-	nodeClientCreator nodeClientCreator
+	driverName          csiDriverName
+	addr                csiAddr
+	nodeV1ClientCreator nodeV1ClientCreator
+	nodeV0ClientCreator nodeV0ClientCreator
 }
 
 var _ csiClient = &csiDriverClient{}
 
-type nodeClientCreator func(driverName string) (
-	nodeClient csipb.NodeClient,
+type nodeV1ClientCreator func(addr csiAddr) (
+	nodeClient csipbv1.NodeClient,
+	closer io.Closer,
+	err error,
+)
+
+type nodeV0ClientCreator func(addr csiAddr) (
+	nodeClient csipbv0.NodeClient,
 	closer io.Closer,
 	err error,
 )
 
-// newNodeClient creates a new NodeClient with the internally used gRPC
+// newV1NodeClient creates a new NodeClient with the internally used gRPC
 // connection set up. It also returns a closer which must to be called to close
 // the gRPC connection when the NodeClient is not used anymore.
-// This is the default implementation for the nodeClientCreator, used in
+// This is the default implementation for the nodeV1ClientCreator, used in
 // newCsiDriverClient.
-func newNodeClient(driverName string) (nodeClient csipb.NodeClient, closer io.Closer, err error) {
+func newV1NodeClient(addr csiAddr) (nodeClient csipbv1.NodeClient, closer io.Closer, err error) {
 	var conn *grpc.ClientConn
-	conn, err = newGrpcConn(driverName)
+	conn, err = newGrpcConn(addr)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	nodeClient = csipb.NewNodeClient(conn)
+	nodeClient = csipbv1.NewNodeClient(conn)
 	return nodeClient, conn, nil
 }
 
-func newCsiDriverClient(driverName string) *csiDriverClient {
-	c := &csiDriverClient{
-		driverName:        driverName,
-		nodeClientCreator: newNodeClient,
+// newV0NodeClient creates a new NodeClient with the internally used gRPC
+// connection set up. It also returns a closer which must to be called to close
+// the gRPC connection when the NodeClient is not used anymore.
+// This is the default implementation for the nodeV1ClientCreator, used in
+// newCsiDriverClient.
+func newV0NodeClient(addr csiAddr) (nodeClient csipbv0.NodeClient, closer io.Closer, err error) {
+	var conn *grpc.ClientConn
+	conn, err = newGrpcConn(addr)
+	if err != nil {
+		return nil, nil, err
 	}
-	return c
+
+	nodeClient = csipbv0.NewNodeClient(conn)
+	return nodeClient, conn, nil
+}
+
+func newCsiDriverClient(driverName csiDriverName) (*csiDriverClient, error) {
+	if driverName == "" {
+		return nil, fmt.Errorf("driver name is empty")
+	}
+
+	addr := fmt.Sprintf(csiAddrTemplate, driverName)
+	requiresV0Client := true
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletPluginsWatcher) {
```

Comment:
it's not obvious to me why client construction would care about how a driver was discovered... is this coupling incidental or inherent?

Reply:
Before the kubelet device plugin registration mechanism existed (when CSI was alpha), the paths to sockets were deterministicly  generated based on driver name instead of discovered. If someone runs with kubelet plugin registration disabled, we fall back that that ancient mechanism way of finding the plugin.

Reply:
ok